neuralpp/experiments/experimental_inference/graph_analysis.py

```diff
@@ -0,0 +1,102 @@
+from typing import Iterable, Set
+
+from neuralpp.inference.graphical_model.representation.factor.product_factor import Factor
+from neuralpp.inference.graphical_model.variable.variable import Variable
+from neuralpp.util import util
+
+
+class Graph:
+    def neighbors(self, node):
+        raise NotImplemented()
+
+
+class FactorGraph(Graph):
+
+    def __init__(self, factors):
+        self.factors = factors
+        self.variable_neighbors = {}
+        # Compute variable neighbors in one pass over factors
+        # to avoid looping over all factors for all variables.
+        for f in factors:
+            for v in f.variables:
+                if self.variable_neighbors.get(v) is None:
+                    self.variable_neighbors[v] = [f]
+                else:
+                    self.variable_neighbors[v] += [f]
+
+    def neighbors(self, node):
+        assert (isinstance(node, Factor) or isinstance(node, Variable))
+        return (node.variables if (isinstance(node, Factor))
+                else self.variable_neighbors.get(node, []))
+
+
+class PartialSpanningTree:
+
+    def __init__(self, graph: Graph, root, stop=lambda _: False):
+        self.graph = graph
+        self.root = root
+        self._children = {}
+        self._parents = {}
+
+    def children(self, node):
+        if not self._children.get(id(node)):
+            neighbors = [n for n in self.graph.neighbors(node) if id(n) not in self._parents]
+            for n in neighbors:
+                self._parents[id(n)] = node
+            self._children[id(node)] = neighbors
+        return self._children[id(node)]
+
+    def parent(self, node):
+        return self._parents.get(id(node), None)
+
+
+def node_variables(node):
+    return node.variables if isinstance(node, Factor) else [node]
+
+
+class FactorPartialSpanningTree(PartialSpanningTree):
+
+    def __init__(self, graph: FactorGraph, root):
+        super().__init__(graph, root)
+
+        def _evaluate(node):
```

Comment:
It is looking to me that `_evaluate` is unnecessary. If you don't do this, when `children` will still be automatically computed when that method is used. This will make the code simpler and potentially improve performance since the calculation only actually happens if needed.